File: internal/store/machines.go
package store

import (
	"slices"
	"time"
)

// ---------------------------------------------------------------------------
// Providers
// ---------------------------------------------------------------------------

// ProviderKind selects the infrastructure a provider rents machines from. It is
// the key the provider registry resolves a factory with, and it is CHECKed in
// the schema, so a kind this build does not know cannot be written and then
// read back as something to act on.
type ProviderKind string

const (
	ProviderProxmox ProviderKind = "proxmox"
	ProviderFake    ProviderKind = "fake"
)

func (k ProviderKind) Valid() bool {
	switch k {
	case ProviderProxmox, ProviderFake:
		return true
	}
	return false
}

// Provider is one place machines can be rented from: the credential, the one
// machine shape it offers, and the ceilings an operator put on it.
//
// The ceilings are the safe default rather than a convenience: MaxMachines is
// zero on a row nobody has configured, which refuses everything. A provider
// that bought machines the moment its credential was accepted would spend money
// on the strength of a connection test.
type Provider struct {
	ID   string       `json:"id"`
	Kind ProviderKind `json:"kind"`
	Name string       `json:"name"`
	// Endpoint, CAPEM and InsecureSkipVerify are how this controller reaches
	// the provider. InsecureSkipVerify is recorded rather than refused because
	// a homelab hypervisor's certificate is usually its own, but it weakens
	// the connection and the validator says so.
	Endpoint           string `json:"endpoint,omitempty"`
	CAPEM              string `json:"ca_pem,omitempty"`
	InsecureSkipVerify bool   `json:"insecure_skip_verify,omitempty"`
	// Settings holds the non-secret answers to the provider's own SettingSpecs
	// -- which node, which storage, which template. It is a map rather than a
	// column per key so that a second provider needs no migration.
	Settings StringMap `json:"settings"`
	// CredentialsEnc is sealed with the instance key and never leaves this
	// process in plaintext. HasSealedSecrets counts it, so a controller that
	// has lost its key refuses to generate a new one rather than starting and
	// failing inside its first call to the hypervisor.
	CredentialsEnc []byte `json:"-"`

	// The one machine shape this provider offers. Every machine it buys is
	// this shape, and a second shape is a second provider row.
	MachineLabels   StringMap   `json:"machine_labels"`
	MachineCapacity int         `json:"machine_capacity"`
	MachineBackend  BackendKind `json:"machine_backend"`
	MachinePlatform Platform    `json:"machine_platform"`
	MachineCPUs     float64     `json:"machine_cpus,omitempty"`
	MachineMemoryMB int64       `json:"machine_memory_mb,omitempty"`
	MachineDiskMB   int64       `json:"machine_disk_mb,omitempty"`

	// PoolSelector narrows which pools this provider may buy for, the way a
	// pool's HostSelector narrows where its runners may be placed. Empty means
	// every pool, which is what one provider and one fleet want.
	PoolSelector StringMap `json:"pool_selector"`
	// MaxMachines is the ceiling on machines this provider may own at once;
	// MaxCreatesInFlight is the ceiling on how many it may be creating. The
	// second exists because a hypervisor cloning four templates at once is
	// slower at all four than it would have been at one.
	MaxMachines        int      `json:"max_machines"`
	MaxCreatesInFlight int      `json:"max_creates_in_flight"`
	IdleTimeout        Duration `json:"idle_timeout"`
	CostPerMachineHour float64  `json:"cost_per_machine_hour,omitempty"`
	Enabled            bool     `json:"enabled"`

	// Paused is the operator's kill switch and PausedUntil the breaker's. They
	// are separate columns because a breaker that expired must not un-press a
	// switch a person pressed.
	Paused              bool       `json:"paused"`
	PausedReason        string     `json:"paused_reason,omitempty"`
	PausedUntil         *time.Time `json:"paused_until,omitempty"`
	ConsecutiveFailures int        `json:"consecutive_failures,omitempty"`
	LastCheckAt         *time.Time `json:"last_check_at,omitempty"`
	LastCheckError      string     `json:"last_check_error,omitempty"`
	LastSweepAt         *time.Time `json:"last_sweep_at,omitempty"`
	CreatedAt           time.Time  `json:"created_at"`
	UpdatedAt           time.Time  `json:"updated_at"`
}

// ---------------------------------------------------------------------------
// Machines
// ---------------------------------------------------------------------------

// MachineState is the lifecycle of one rented machine, from a row that names
// nothing to a host that runs jobs.
type MachineState string

const (
	MachinePlanned       MachineState = "planned"       // the reservation exists; nothing has been created
	MachineCreating      MachineState = "creating"      // identity persisted; a create may or may not have happened
	MachineStarting      MachineState = "starting"      // the resource exists and is being powered on
	MachineBootstrapping MachineState = "bootstrapping" // the guest is up; the agent is being installed
	MachineEnrolling     MachineState = "enrolling"     // the agent has its credential and has not joined
	MachineReady         MachineState = "ready"         // a host row exists and is linked
	MachineDraining      MachineState = "draining"      // cordoned; waiting for its runners to finish
	MachineDeleting      MachineState = "deleting"      // a delete was issued and is not confirmed
	MachineDeleted       MachineState = "deleted"       // the provider confirmed the resource is gone
	MachineFailed        MachineState = "failed"        // we gave up on the lifecycle; the resource may still exist
	MachineQuarantined   MachineState = "quarantined"   // ownership could not be proved; only an operator moves it
)

func (s MachineState) Valid() bool {
	switch s {
	case MachinePlanned, MachineCreating, MachineStarting, MachineBootstrapping,
		MachineEnrolling, MachineReady, MachineDraining, MachineDeleting,
		MachineDeleted, MachineFailed, MachineQuarantined:
		return true
	}
	return false
}

// Terminal is deleted alone. Failed and quarantined still need attention, and a
// failed machine may still have a resource behind it.
func (s MachineState) Terminal() bool { return s == MachineDeleted }

// Pending is a machine the fleet is waiting on: it has no host yet, so its
// capacity is a reservation rather than a fact.
func (s MachineState) Pending() bool {
	switch s {
	case MachinePlanned, MachineCreating, MachineStarting, MachineBootstrapping, MachineEnrolling:
		return true
	}
	return false
}

// Offers is a machine whose capacity counts towards satisfying demand.
func (s MachineState) Offers() bool { return s.Pending() || s == MachineReady }

// MachineOpKind is the one kind of operation that may be in flight against a
// machine. The empty string is "none", and the schema CHECKs the set.
type MachineOpKind string

const (
	MachineOpNone      MachineOpKind = ""
	MachineOpCreate    MachineOpKind = "create"
	MachineOpStart     MachineOpKind = "start"
	MachineOpStop      MachineOpKind = "stop"
	MachineOpBootstrap MachineOpKind = "bootstrap"
	MachineOpDelete    MachineOpKind = "delete"
)

func (k MachineOpKind) Valid() bool {
	switch k {
	case MachineOpNone, MachineOpCreate, MachineOpStart, MachineOpStop, MachineOpBootstrap, MachineOpDelete:
		return true
	}
	return false
}

// MachineOperation is one pass's claim on a machine's next step.
//
// Timeout is required, and ClaimMachineOperation refuses a claim without one:
// a claim with no deadline is a claim a controller that dies mid-operation
// holds for ever, and the machine behind it would never be looked at again.
type MachineOperation struct {
	// ID is minted per attempt when it is empty, and is what a log line and an
	// operator quote. Holder is the controller taking the step, so a machine
	// stuck mid-create names the controller to go and look at.
	ID      string
	Kind    MachineOpKind
	Holder  string
	Timeout time.Duration
}

// MachineOpOutcome is how an operation ended, which decides what happens to the
// retry accounting rather than to the state -- the state is TransitionMachine's.
type MachineOpOutcome string

const (
	// MachineOpSucceeded clears the provider's complaint and the backoff.
	MachineOpSucceeded MachineOpOutcome = "succeeded"
	// MachineOpFailed counts an attempt and sets the next one.
	MachineOpFailed MachineOpOutcome = "failed"
	// MachineOpReleased gives the claim back without a verdict: a guard that
	// refused after the claim was taken, or a controller shutting down. It
	// must not count an attempt, because nothing about the machine went wrong
	// and counting it would walk the backoff up towards giving up.
	MachineOpReleased MachineOpOutcome = "released"
)

func (o MachineOpOutcome) Valid() bool {
	switch o {
	case MachineOpSucceeded, MachineOpFailed, MachineOpReleased:
		return true
	}
	return false
}

// MachineErrorSource says which of the two independent systems complained. They
// have a column each for the reason 0022 gives about runner cleanup: a success
// on one must not erase the other's complaint, and an operator reading one
// error has to know which half to go and look at.
type MachineErrorSource string

const (
	// MachineErrorProvider is the hypervisor or cloud API.
	MachineErrorProvider MachineErrorSource = "provider"
	// MachineErrorBootstrap is everything inside the guest: the agent install,
	// the enrolment, the handshake back.
	MachineErrorBootstrap MachineErrorSource = "bootstrap"
)

func (s MachineErrorSource) Valid() bool {
	return s == MachineErrorProvider || s == MachineErrorBootstrap
}

// Machine is one rented machine: the row that exists before the resource does,
// and outlives it by exactly as long as it takes to prove the resource is gone.
type Machine struct {
	ID         string       `json:"id"`
	ProviderID string       `json:"provider_id"`
	Name       string       `json:"name"`
	State      MachineState `json:"state"`
	Message    string       `json:"message,omitempty"`
	// PoolID records which pool's unmet demand asked for this machine. It is
	// not ownership -- a machine serves every pool its labels match -- and it
	// exists so the machine page can answer "why does this exist".
	PoolID string `json:"pool_id,omitempty"`

	// OwnerControllerID and OwnerFingerprint are written before the first call
	// that can create anything. The fingerprint is tamper-evidence rather than
	// a credential: anybody with audit rights on the hypervisor can read it,
	// and its job is to tell "the VM I created at 143" apart from "a VM
	// somebody else created at 143 after mine was destroyed".
	OwnerControllerID string `json:"owner_controller_id,omitempty"`
	OwnerFingerprint  string `json:"owner_fingerprint,omitempty"`
	// ResourceZone and ResourceID are the provider's own identity for this
	// machine -- a node and a VM id. Written once, never moved.
	ResourceZone   string    `json:"resource_zone,omitempty"`
	ResourceID     string    `json:"resource_id,omitempty"`
	ResourceDetail StringMap `json:"resource_detail,omitempty"`
	Address        string    `json:"address,omitempty"`
	// OwnershipVerifiedAt is the last time a fresh observation agreed that this
	// resource is ours. OwnershipError is what disagreed. A delete needs the
	// first and the absence of the second.
	OwnershipVerifiedAt *time.Time `json:"ownership_verified_at,omitempty"`
	OwnershipError      string     `json:"ownership_error,omitempty"`

	// HostID is the enrolment link, and the only thing that grants this
	// controller deletion authority over a host. It is written by
	// LinkMachineHost after a token this row minted was redeemed, and by
	// nothing else -- there is no route that attaches a machine to a host an
	// operator already had.
	HostID      string    `json:"host_id,omitempty"`
	JoinTokenID string    `json:"join_token_id,omitempty"`
	Capacity    int       `json:"capacity,omitempty"`
	Labels      StringMap `json:"labels,omitempty"`

	// The operation in flight, if any. OpHandle is the provider's own handle
	// for the call -- a task id -- and is what recovery asks about after a
	// restart. OpOutcomeUnknown marks a call that went out and was never
	// answered, which is the one thing a retry must not treat as "it failed".
	OpID             string        `json:"op_id,omitempty"`
	OpKind           MachineOpKind `json:"op_kind,omitempty"`
	OpHandle         string        `json:"op_handle,omitempty"`
	OpHolder         string        `json:"op_holder,omitempty"`
	OpStartedAt      *time.Time    `json:"op_started_at,omitempty"`
	OpDeadlineAt     *time.Time    `json:"op_deadline_at,omitempty"`
	OpOutcomeUnknown bool          `json:"op_outcome_unknown,omitempty"`

	Attempts       int        `json:"attempts,omitempty"`
	NextAttemptAt  *time.Time `json:"next_attempt_at,omitempty"`
	ProviderError  string     `json:"provider_error,omitempty"`
	BootstrapError string     `json:"bootstrap_error,omitempty"`

	CreatedAt            time.Time  `json:"created_at"`
	UpdatedAt            time.Time  `json:"updated_at"`
	ReservationExpiresAt *time.Time `json:"reservation_expires_at,omitempty"`
	CreateStartedAt      *time.Time `json:"create_started_at,omitempty"`
	CreatedOKAt          *time.Time `json:"created_ok_at,omitempty"`
	StartedAt            *time.Time `json:"started_at,omitempty"`
	BootstrappedAt       *time.Time `json:"bootstrapped_at,omitempty"`
	EnrolledAt           *time.Time `json:"enrolled_at,omitempty"`
	ReadyAt              *time.Time `json:"ready_at,omitempty"`
	IdleSince            *time.Time `json:"idle_since,omitempty"`
	DrainingAt           *time.Time `json:"draining_at,omitempty"`
	DeleteStartedAt      *time.Time `json:"delete_started_at,omitempty"`
	// DeletedAt is when the provider confirmed the resource was gone, by an
	// inspect that could not find it. A 200 from the delete call is not that,
	// which is why only ConfirmMachineDeleted writes this.
	DeletedAt *time.Time `json:"deleted_at,omitempty"`
}

// Owns reports whether this row believes a resource exists in the provider. It
// is deliberately NOT a state test: a failed machine whose VM was created still
// costs money and still counts against max_machines, and a planned one costs
// nothing. The SQL half of this predicate lives in CountOwnedMachines, and a
// test holds the two definitions equal -- live_rows_test.go's reason.
func (m *Machine) Owns() bool { return m.ResourceID != "" && m.DeletedAt == nil }

// validMachineTransitions is the allow-list Store.TransitionMachine enforces. It
// is here rather than in the reconciler for the same reason the runner one is: a
// pass that has just lost a race, or a provider that answered oddly, must not be
// able to talk a machine into a state that makes the fleet's accounting wrong.
//
// The absences are the design:
//   - nothing returns to planned: you never go back to "nothing exists";
//   - deleting leads only to deleted or quarantined, because a delete that was
//     issued cannot be un-issued; if we cannot tell, an operator is asked;
//   - quarantined never returns to ready: a machine whose ownership we could not
//     prove does not silently rejoin the fleet;
//   - failed may still reach deleting, because a later sweep can find a resource
//     we believed never existed;
//   - draining returns to ready, so demand coming back cancels a drain rather
//     than paying for a new machine.
var validMachineTransitions = map[MachineState][]MachineState{
	MachinePlanned:       {MachineCreating, MachineFailed, MachineDeleted, MachineQuarantined},
	MachineCreating:      {MachineStarting, MachineBootstrapping, MachineDeleting, MachineFailed, MachineQuarantined},
	MachineStarting:      {MachineBootstrapping, MachineDeleting, MachineFailed, MachineQuarantined},
	MachineBootstrapping: {MachineEnrolling, MachineDeleting, MachineFailed, MachineQuarantined},
	MachineEnrolling:     {MachineReady, MachineDeleting, MachineFailed, MachineQuarantined},
	MachineReady:         {MachineDraining, MachineDeleting, MachineFailed, MachineQuarantined},
	MachineDraining:      {MachineReady, MachineDeleting, MachineFailed, MachineQuarantined},
	MachineDeleting:      {MachineDeleted, MachineQuarantined},
	MachineDeleted:       {},
	MachineFailed:        {MachineDeleting, MachineDeleted, MachineQuarantined},
	MachineQuarantined:   {MachineDeleting, MachineDeleted},
}

// CanTransitionMachine reports whether from -> to is a legal machine state
// change. A self-transition is always legal, because a pass re-reporting the
// state a machine is already in is not an error and must not become one.
func CanTransitionMachine(from, to MachineState) bool {
	if from == to {
		return true
	}
	return slices.Contains(validMachineTransitions[from], to)
}

File: web/src/lib/providers/machines.ts
/**
 * What a rented machine is doing, worked out from the row alone.
 *
 * Pure, and in a `.ts` rather than a component, for the reason
 * `insights/signals.ts` is: the arithmetic behind "this one has been
 * bootstrapping for eleven minutes" and "these three need a person" is what
 * goes wrong, and it is testable here without a browser.
 *
 * Nothing here reads a clock of its own. `now` is passed in, so a test can say
 * what time it is and a component can hand it the same tick it renders with.
 */
import type { Machine, MachineState, Provider } from '../api/types';

/**
 * The lifecycle as the band draws it: the states a machine passes through on
 * its way to being a host. Deleting, deleted, failed and quarantined are exits
 * from the flow rather than steps of it, and are counted separately.
 */
export const MACHINE_FLOW: readonly MachineState[] = [
  'planned',
  'creating',
  'starting',
  'bootstrapping',
  'enrolling',
  'ready',
  'draining',
];

/** The states a machine is in while the fleet is still waiting for it. */
const PENDING: ReadonlySet<MachineState> = new Set([
  'planned',
  'creating',
  'starting',
  'bootstrapping',
  'enrolling',
]);

/** The phases a timeline entry can carry, as a sentence for the page. */
const PHASES: Readonly<Record<string, string>> = {
  planned: 'Planned',
  creating: 'Creating the machine',
  created: 'Machine created',
  starting: 'Powering on',
  bootstrapped: 'Agent installed',
  enrolled: 'Enrolled with the controller',
  ready: 'Ready for work',
  draining: 'Draining',
  deleting: 'Deleting',
  deleted: 'Resource confirmed gone',
};

/**
 * What a phase is called on the page. Two consecutive rows both reading
 * "Creating" say nothing, and those rows are exactly the diagnosis: a clone
 * that never finished is a problem at the hypervisor, and a machine that
 * cloned and never enrolled is a problem inside the guest.
 */
export function phaseLabel(phase: string | undefined): string {
  if (!phase) return 'Unknown';
  return PHASES[phase] ?? phase.charAt(0).toUpperCase() + phase.slice(1);
}

/**
 * The state a timeline phase belongs to, so a row is drawn in the colour the
 * rest of the product gives that part of a machine's life. The timeline has
 * more entries than there are states -- "created" and "creating" are two marks
 * on one state -- because what an operator wants to know is which half of a
 * slow create is slow.
 */
const PHASE_STATES: Readonly<Record<string, MachineState>> = {
  planned: 'planned',
  creating: 'creating',
  created: 'creating',
  starting: 'starting',
  bootstrapped: 'bootstrapping',
  enrolled: 'enrolling',
  ready: 'ready',
  draining: 'draining',
  deleting: 'deleting',
  deleted: 'deleted',
};

export function phaseState(phase: string | undefined): MachineState | undefined {
  return phase ? PHASE_STATES[phase] : undefined;
}

export interface MachineSignals {
  /** Whether the fleet is still waiting for this machine to become a host. */
  pending: boolean;
  /** Whether it holds a resource somebody is being billed for. */
  owns: boolean;
  /** How long it has been in its current state, or null when that is unknown. */
  elapsedMs: number | null;
  /** The operation in flight, as a word, or '' when none is. */
  operation: string;
  /**
   * Whether an operator has to look at this one. It is deliberately not "the
   * state is failed": a machine whose ownership nothing could confirm, and one
   * whose last call went out and was never answered, are both running
   * hardware nothing will tidy up on its own.
   */
  needsReview: boolean;
  /** Why, in one sentence, or '' when it does not. */
  reviewReason: string;
  /** The row's own reading of whether a delete would be safe. */
  safeToDelete: boolean;
  safeToDeleteWhy: string;
}

export function machineSignals(machine: Machine, now: number = Date.now()): MachineSignals {
  const since = machine.updated_at ? Date.parse(machine.updated_at) : Number.NaN;
  const state = machine.state;
  return {
    pending: state !== undefined && PENDING.has(state),
    owns: Boolean(machine.resource_id) && !machine.deleted_at,
    elapsedMs: Number.isFinite(since) ? Math.max(0, now - since) : null,
    operation: machine.operation ?? '',
    ...review(machine),
    safeToDelete: machine.safe_to_delete === true,
    safeToDeleteWhy: machine.safe_to_delete_why ?? '',
  };
}

function review(machine: Machine): { needsReview: boolean; reviewReason: string } {
  if (machine.state === 'quarantined') {
    return {
      needsReview: true,
      reviewReason:
        machine.ownership_error ||
        'Nothing could prove this resource is ours, so nothing here will act on it.',
    };
  }
  if (machine.ownership_error) return { needsReview: true, reviewReason: machine.ownership_error };
  if (machine.state === 'failed') {
    return {
      needsReview: true,
      reviewReason:
        machine.provider_error ||
        machine.bootstrap_error ||
        'This machine never became a host. Its resource may still exist.',
    };
  }
  if (machine.outcome_unknown) {
    return {
      needsReview: true,
      reviewReason: 'A call went out and was never answered, so what happened is not known yet.',
    };
  }
  return { needsReview: false, reviewReason: '' };
}

/** Every machine an operator has to look at, in the order they were listed. */
export function machinesNeedingReview(machines: readonly Machine[]): Machine[] {
  return machines.filter((m) => review(m).needsReview);
}

export interface LifecycleStep {
  state: MachineState;
  count: number;
}

/**
 * How many machines are at each step of the flow.
 *
 * Deleted machines are left out rather than counted as zero: a fleet that has
 * been running for a month has hundreds of them, and a band whose largest
 * number is the history is a band nobody reads.
 */
export function lifecycleCounts(machines: readonly Machine[]): LifecycleStep[] {
  const counts = new Map<MachineState, number>();
  for (const state of MACHINE_FLOW) counts.set(state, 0);
  for (const machine of machines) {
    const state = machine.state;
    if (state === undefined || !counts.has(state)) continue;
    counts.set(state, (counts.get(state) ?? 0) + 1);
  }
  return MACHINE_FLOW.map((state) => ({ state, count: counts.get(state) ?? 0 }));
}

/**
 * What a machine has cost so far, or null when its provider has no opinion.
 *
 * A provider with no price is the common case -- a hypervisor in a cupboard
 * has no hourly rate -- and inventing one would be worse than saying nothing.
 */
export function machineCost(
  machine: Machine,
  provider: Pick<Provider, 'cost_per_machine_hour'> | undefined,
  now: number = Date.now(),
): number | null {
  const rate = provider?.cost_per_machine_hour ?? 0;
  if (rate <= 0) return null;
  const from = machine.created_at ? Date.parse(machine.created_at) : Number.NaN;
  if (!Number.isFinite(from)) return null;
  const to = machine.deleted_at ? Date.parse(machine.deleted_at) : now;
  if (!Number.isFinite(to)) return null;
  return (Math.max(0, to - from) / 3_600_000) * rate;
}

/**
 * A cost as a figure. Two decimals, because that is what an invoice has, and
 * an empty string rather than "0.00" when there is no price to report -- a
 * zero reads as free, which is a different claim from "nobody said".
 */
export function formatCost(value: number | null): string {
  return value === null ? '' : value.toFixed(2);
}
